File: app/src/main/java/com/hundaol/ethiopic/views/CalendarViewGestureDetector.java
package com.hundaol.ethiopic.views;

import android.os.Handler;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.widget.Scroller;

import com.hundaol.ethiopic.domain.DateModel;

/**
 * Created by john.pirie on 2017-04-29.
 */

public class CalendarViewGestureDetector {

    private final GestureDetector gestureDetector;

    public CalendarViewGestureDetector(CalendarView view, DateModel dateModel) {
        this.gestureDetector = new GestureDetector(view.getContext(), new GestureDetector.OnGestureListener() {
            @Override
            public boolean onDown(MotionEvent e) {
                return true;
            }

            @Override
            public void onShowPress(MotionEvent e) {

            }

            @Override
            public boolean onSingleTapUp(MotionEvent e) {
                return false;
            }

            @Override
            public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
                float cellWidth = view.getViewModel().getCellWidth();
                float jdv = view.getViewModel().getJdv();
                jdv = jdv + 7.0f * distanceY / cellWidth;
                dateModel.setJdv(jdv);
                return true;
            }

            @Override
            public void onLongPress(MotionEvent e) {

            }

            @Override
            public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
                Handler handler = new Handler();
                Scroller scroller = new Scroller(view.getContext());
                scroller.setFriction(4 * 0.015f);
                scroller.fling(0, 0, 0, (int) velocityY, Integer.MIN_VALUE, Integer.MAX_VALUE, Integer.MIN_VALUE, Integer.MAX_VALUE);
                final float anchorJdv = view.getViewModel().getJdv();
                handler.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        if (scroller.computeScrollOffset()) {
                            dateModel.setJdv(anchorJdv - scroller.getCurrY() * 7 / view.getViewModel().getCellWidth());
                            handler.postDelayed(this, 100);
                        }
                    }
                }, 100);
                return true;
            }
        });

        view.setOnTouchListener((v, event) -> gestureDetector.onTouchEvent(event));
    }
}

File: app/src/main/java/com/hundaol/ethiopic/views/calendar/CalendarViewGestureDetector.kt
package com.hundaol.ethiopic.views.calendar

import android.os.Handler
import android.view.GestureDetector
import android.view.MotionEvent
import android.widget.Scroller
import com.hundaol.ethiopic.App
import com.hundaol.ethiopic.views.calendar.CalendarView

/**
 * Created by john.pirie on 2017-04-29.
 */

class CalendarViewGestureDetector(val view: CalendarView) {

    private val gestureDetector: GestureDetector

    init {
        this.gestureDetector = GestureDetector(view.context, object : GestureDetector.OnGestureListener {
            override fun onDown(e: MotionEvent): Boolean {
                return true
            }

            override fun onShowPress(e: MotionEvent) {

            }

            override fun onSingleTapUp(e: MotionEvent): Boolean {
                return false
            }

            override fun onScroll(e1: MotionEvent, e2: MotionEvent, distanceX: Float, distanceY: Float): Boolean {
                val cellWidth = view.viewModel.cellWidth
                var jdv = view.dateModel.jdv
                jdv = jdv + 7.0f * distanceY / cellWidth
                App.Companion.setJdv(jdv)
                return true
            }

            override fun onLongPress(e: MotionEvent) {}

            override fun onFling(e1: MotionEvent, e2: MotionEvent, velocityX: Float, velocityY: Float): Boolean {
                val handler = Handler()
                val scroller = Scroller(view.context)
                scroller.setFriction(4 * 0.015f)
                scroller.fling(0, 0, 0, velocityY.toInt(), Integer.MIN_VALUE, Integer.MAX_VALUE, Integer.MIN_VALUE, Integer.MAX_VALUE)
                val anchorJdv = view.dateModel.jdv
                handler.postDelayed(object : Runnable {
                    override fun run() {
                        if (scroller.computeScrollOffset()) {
                            App.Companion.setJdv(anchorJdv - scroller.currY * 7 / view.viewModel.cellWidth)
                            handler.postDelayed(this, 100)
                        }
                    }
                }, 100)
                return true
            }
        })

        view.setOnTouchListener { v, event -> gestureDetector.onTouchEvent(event) }
    }
}
